File: bluetooth_printer/lib/src/bluetooth_printer.dart
import 'package:flet/flet.dart';
import 'package:flutter/material.dart';
import 'package:my_bluetooth_printer/my_bluetooth_printer_util.dart';
import 'package:my_bluetooth_printer/my_bluetooth_printer_mixin.dart';

class BluetoothPrinter extends StatefulWidget {
  final Control? parent;
  final Control control;
  final FletControlBackend backend;

  const BluetoothPrinter(
      {super.key, this.parent, required this.control, required this.backend});
  @override
  _Bluetooth createState() => _Bluetooth();
}

class _Bluetooth extends State<BluetoothPrinter> with MyBluetoothPrinterMixin<BluetoothPrinter>{
  late TextEditingController _controller;
  String _string = "";
  bool _set_channel_name = true;
  // bool _start_scan = false;
  // bool _isConnect = false;
  // late List<String> _bluetooth_list;
  // late int _connect_message = 0;
  MyBluetoothPrinterUtil myBluetoothUtil = MyBluetoothPrinterUtil();

  @override
  Widget build(BuildContext context) {
    var channel_name = widget.control.attrString("channel_name");
    if (_set_channel_name && channel_name != null) {
      myBluetoothUtil.setMessageChannel(channel_name, listenerAndroidHandle);
      myBluetoothUtil.sendChannelName("channelName", channel_name);
      setState(() {
        _set_channel_name = false;
      });
    }
    var message_tag = widget.control.attrString("message_tag");
    var start_scan = widget.control.attrBool("start_scan")!;
    var stop_scan = widget.control.attrBool("stop_scan")!;
    var start_connect = widget.control.attrBool("start_connect")!;
    var close_connect = widget.control.attrBool("close_connect")!;
    var start_send = widget.control.attrBool("start_send")!;
    var address = widget.control.attrString("address");
    var send_data = widget.control.attrString("send_data");
    var bluetooth_list = widget.control.attrString("bluetooth_list");
    var connect_message = widget.control.attrString("connect_message");
    var connect_response = widget.control.attrString("connect_response");
    var data_response = widget.control.attrString("data_response");
    var scan_message = widget.control.attrString("scan_message");
    
    if (start_scan) {     // 开始扫描标志为True时向Android端发送对应信息
      myBluetoothUtil.sendMessageToAndroid("startScan", true);
      setState(() {
        /*
          发送完消息后将开始扫描标志置为“false”（目前只掌握更新字符串属性方式，
          不过影响不大python控件那边布尔型属性可以接受字符串）
          下方多个标志一样的处理方式
        */
        widget.backend.updateControlState(widget.control.id, {"start_scan": "false"});
      });
    } 
    if (stop_scan)  {
      myBluetoothUtil.sendMessageToAndroid("stopScan", true);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"stop_scan": "false"});
      });
    }
    if (start_connect && address != "") {    // 连接设备标志
      myBluetoothUtil.sendMessageToAndroid("startConnect", address);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"start_connect": "false"});
      });
    }
    if (close_connect) {    // 关闭设备连接标志
      myBluetoothUtil.sendMessageToAndroid("closeConnect", true);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"close_connect": "false"});
      });
    }
    if (start_send) {
      myBluetoothUtil.sendMessageToAndroid("startSend", send_data);
      setState(() {
        widget.backend.updateControlState(widget.control.id, {"start_send": "false"});
      });
    }
    if (_string != "") {
      _controller.text = _string;
    }
    Widget bluetooth_control = TextField(
        controller: _controller,
        decoration: const InputDecoration(
          border: InputBorder.none, //无边框
        )
    );
    return constrainedControl(
        context, bluetooth_control, widget.parent, widget.control);
  }

  @override
  Future<void> listenerAndroidHandle(dynamic message) async {
    // TODO: implement listenerAndroidHandle
    var connectResponse = message["connectResponse"];
    if (connectResponse != null) {
      // 连接回传数据
      widget.backend.updateControlState(widget.control.id, {"connect_response": connectResponse});
      String eventString = connectResponse;
      widget.backend.updateControlState(widget.control.id, {"message_tag": "connect_response"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
    var dataResponse = message["dataResponse"];
    if (dataResponse != null) {
      // 数据回传信息
      widget.backend.updateControlState(widget.control.id, {"data_response": dataResponse});
      String eventString = dataResponse;
      widget.backend.updateControlState(widget.control.id, {"message_tag": "data_response"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
    // 设备列表
    var bluetoothList = message["bluetooth_list"];
    if (bluetoothList != null) {
      String eventString = "";
      for (String bluetooth in bluetoothList) {
        eventString += "$bluetooth&";
      }
      widget.backend.updateControlState(widget.control.id, {"bluetooth_list": eventString});
      widget.backend.updateControlState(widget.control.id, {"message_tag": "bluetooth_list"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
    // 连接结果（连接成功、连接失败、断开连接）消息样式为   连接成功>>>（设备名称）
    var connectMessage = message["connectMessage"];
    if (connectMessage != null) {
      widget.backend.updateControlState(widget.control.id, {"connect_message": connectMessage});
      String eventString = connectMessage;
      widget.backend.updateControlState(widget.control.id, {"message_tag": "connect_message"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
    var scanMessage = message["scanMessage"];
    if (scanMessage != null) {
      widget.backend.updateControlState(widget.control.id, {"scan_message": scanMessage});
      String eventString = scanMessage;
      widget.backend.updateControlState(widget.control.id, {"message_tag": "scan_message"});
      widget.backend.triggerControlEvent(widget.control.id, "on_listener", eventString);
      setState(() {
        _string = eventString;
      });
    }
  }
}

File: python/controls/bluetooth_printer.py
from flet_core.constrained_control import ConstrainedControl
from typing import Optional

class BluetoothPrinter(ConstrainedControl):
    """
    蓝牙打印机控件
    """
    def __init__(
        self,
        address: Optional[str] = "",
        send_data: Optional[str] = "",
        start_scan: Optional[bool] = False,
        stop_scan: Optional[bool] = False,
        start_connect: Optional[bool] = False,
        close_connect: Optional[bool] = False,
        start_send: Optional[bool] = False,
        bluetooth_list: Optional[str] = "",
        connect_message: Optional[str] = "",
        connect_response: Optional[str] = "",
        data_response: Optional[str] = "",
        scan_message: Optional[str] = "",
        message_tag: Optional[str] = "",
        on_listener = None,
    ):
        ConstrainedControl.__init__(self)
        self.address = address
        self.send_data = send_data
        self.start_scan = start_scan
        self.stop_scan = stop_scan
        self.start_connect = start_connect
        self.close_connect = close_connect
        self.start_send = start_send
        self.bluetooth_list = bluetooth_list
        self.connect_message = connect_message
        self.connect_response = connect_response
        self.data_response = data_response
        self.scan_message = scan_message
        self.message_tag = message_tag
        self.on_listener = on_listener
        self.channel_name = str(id(self))
    
    def _get_control_name(self) -> str:
        return "bluetooth_printer"
    
    @property
    def address(self):
        return self._get_attr("address")

    @address.setter
    def address(self, value):
        self._set_attr("address", value)
    
    @property
    def send_data(self):
        return self._get_attr("send_data")

    @send_data.setter
    def send_data(self, value):
        self._set_attr("send_data", value)

    @property
    def start_scan(self):
        return self._get_attr("start_scan")

    @start_scan.setter
    def start_scan(self, value):
        self._set_attr("start_scan", value)

    @property
    def stop_scan(self):
        return self._get_attr("stop_scan")

    @stop_scan.setter
    def stop_scan(self, value):
        self._set_attr("stop_scan", value)
    
    @property
    def start_connect(self):
        return self._get_attr("start_connect")

    @start_connect.setter
    def start_connect(self, value):
        self._set_attr("start_connect", value)

    @property
    def close_connect(self):
        return self._get_attr("close_connect")

    @close_connect.setter
    def close_connect(self, value):
        self._set_attr("close_connect", value)

    @property
    def start_send(self):
        return self._get_attr("start_send")

    @start_send.setter
    def start_send(self, value):
        self._set_attr("start_send", value)

    @property
    def bluetooth_list(self):
        return self._get_attr("bluetooth_list")

    @bluetooth_list.setter
    def bluetooth_list(self, value):
        self._set_attr("bluetooth_list", value)

    @property
    def connect_message(self):
        return self._get_attr("connect_message")

    @connect_message.setter
    def connect_message(self, value):
        self._set_attr("connect_message", value)

    @property
    def connect_response(self):
        return self._get_attr("connect_response")

    @connect_response.setter
    def connect_response(self, value):
        self._set_attr("connect_response", value)

    @property
    def data_response(self):
        return self._get_attr("data_response")

    @data_response.setter
    def data_response(self, value):
        self._set_attr("data_response", value)

    @property
    def scan_message(self):
        return self._get_attr("scan_message")

    @scan_message.setter
    def scan_message(self, value):
        self._set_attr("scan_message", value)

    @property
    def message_tag(self):
        return self._get_attr("message_tag")

    @message_tag.setter
    def message_tag(self, value):
        self._set_attr("message_tag", value)
    
    @property
    def channel_name(self):
        return self._get_attr("channel_name")

    @channel_name.setter
    def channel_name(self, value):
        self._set_attr("channel_name", value)
    
    @property
    def on_listener(self):
        return self._get_event_handler("on_listener")

    @on_listener.setter
    def on_listener(self, handler):
        self._add_event_handler("on_listener", handler)
    
    def scan(self):
        self._set_attr("start_scan", True)
    
    def stop(self):
        self._set_attr("stop_scan", True)
    
    def connect(self, address):
        self._set_attr("address", address)
        self._set_attr("start_connect", True)
    
    def close(self):
        self._set_attr("close_connect", True)
    
    def send(self, send_data):
        self._set_attr("send_data", send_data)
        self._set_attr("start_send", True)